sdks/python/apache_beam/transforms/combinefn_lifecycle_test.py

```diff
@@ -0,0 +1,147 @@
+    # properly after pipeline execution.
+    self.instances.add(self)
+    self._setup_called = True
+
+  def create_accumulator(self):
+    assert self._setup_called, 'setup should have been called'
+    assert not self._teardown_called, 'teardown should not have been called'
+    self._accumulators_created += 1
+    return 0
+
+  def add_input(self, mutable_accumulator, element):
+    assert self._setup_called, 'setup should have been called'
+    assert self._accumulators_created > 0, \
+        'create_accumulator should have been called'
+    assert not self._teardown_called, 'teardown should not have been called'
+    mutable_accumulator += element
+    return mutable_accumulator
+
+  def add_inputs(self, mutable_accumulator, elements):
+    return self.add_input(mutable_accumulator, sum(elements))
+
+  def merge_accumulators(self, accumulators):
+    assert self._setup_called, 'setup should have been called'
+    assert not self._teardown_called, 'teardown should not have been called'
+    return sum(accumulators)
+
+  def extract_output(self, accumulator):
+    assert self._setup_called, 'setup should have been called'
+    assert not self._teardown_called, 'teardown should not have been called'
+    return accumulator
+
+  def teardown(self):
+    assert self._setup_called, 'setup should have been called'
+    assert not self._teardown_called, 'teardown should not be called twice'
+    self._teardown_called = True
+
+
+class BaseCombineFnLifecycleTest(unittest.TestCase):
+  def start(self, pipeline, lift_combiners=True):
+    with pipeline as p:
+      pcoll = p | 'Start' >> beam.Create(range(5))
+
+      # Certain triggers, such as AfterCount, are incompatible with combiner
+      # lifting. We can use that fact to prevent combiners from being lifted.
+      if not lift_combiners:
+        pcoll |= beam.WindowInto(
+            window.GlobalWindows(),
+            trigger=trigger.AfterCount(5),
+            accumulation_mode=trigger.AccumulationMode.DISCARDING)
+
+      pcoll |= 'Do' >> beam.CombineGlobally(CallSequenceEnforcingCombineFn())
+      assert_that(pcoll, equal_to([10]))
+
+    # Ensure that _teardown_called equals True for all CombineFns.
+    for instance in CallSequenceEnforcingCombineFn.instances:
+      self.assertTrue(instance._teardown_called)
+
+
+@attr('ValidatesRunner')
+class CombineFnLifecycleTest(BaseCombineFnLifecycleTest):
+  def setUp(self):
+    self.pipeline = TestPipeline(is_integration_test=True)
+    options = self.pipeline.get_pipeline_options()
+    standard_options = options.view_as(StandardOptions)
+    experiments = options.view_as(DebugOptions).experiments or []
+
+    if 'DataflowRunner' in standard_options.runner and \
+       not standard_options.streaming and \
+       'beam_fn_api' not in experiments and 'use_runner_v2' not in experiments:
+      self.skipTest(
+          'Non-portable Dataflow batch worker does not support '
```

Comment:
Should we make  non-portable Dataflow runner detect usage of combiner initialization and alert the user that this functionality is unsupported?
cc: @robertwb

Reply:
Yes, it's probably a good idea. 
Should we raise an exception and exit the program abnormally if user-provided setup and teardown are detected, or just inform the user that those methods won't be called?

Reply:
On BEAM-3736 the recommendation was to reject the job. It is more problematic with Java SDK. AFAIK with Python SDK, Dataflow is the only non-portable runner (not including direct runner).

Reply:
I've added a code that rejects the job and provides clear information why it's rejected.